js/editor-controller.js

```diff
@@ -0,0 +1,229 @@
+/**
+ * The Editor can be controlled by remote hosts that wrap the editor with
+ * an iframe or in some kind of webview.
+ */
+'use strict';
+
+/**
+ * @returns An object with the fs wrapper.
+ */
+var EditorController = function() {
+    var _returnObj = {};
+
+    var _editorActions = null;
+    var _controllerHost = null;
+
+    /** Constants used for iframe messaging */
+    var CONTROLLER_MESSAGING = Object.freeze({
+        // Embed editor host type
+        type: 'pyeditor',
+        // Embed editor messaging actions
+        actions: {
+            workspacesync: 'workspacesync',
+            workspacesave: 'workspacesave',
+            workspaceloaded: 'workspaceloaded',
+            importproject: 'importproject',
+            loadhex: 'loadhex',
+            loadfile: 'loadfile',
+            mobilemode: 'mobilemode',
+            savefile: 'savefile',
+            flashhex: 'flashhex',
+        }
+    });
+
+    /**
+     * JS debounce for a given amount of time. If the same function is called
+     * before the timeout the timer resets.
+     *
+     * @param {function} callback - Function to debounce. 
+     * @param {number} wait - Time to wait without calling this debounce to
+     *      finally execute the callback.
+     */
+    function debounce(callback, wait) {
+        var timeout = null;
+        return function() {
+            var args = arguments;
+            var next = function() {
+                return callback.apply(this, args);
+            };
+            clearTimeout(timeout);
+            timeout = setTimeout(next, wait);
+        }
+    }
+
+    /**
+     * Sending a postMessage to the host controller.
+     */
+    function hostPostMessage() {
+        if (_controllerHost) {
+            _controllerHost.postMessage.apply(null, arguments);
+        } else {
+            console.error('Trying to postMessage to undefined host controller:');
+            console.error(arguments);
+        }
+    }
+
+    /**
+     * Send data to the host controller to save a file into the device.
+     *
+     * @param {string} filename - File name for the file to save to the device.
+     * @param {string} fileStr - File content to save to the device.
+     */
+    function hostSaveFile(filename, fileStr) {
+        hostPostMessage({
+            type: CONTROLLER_MESSAGING.type,
+            action: CONTROLLER_MESSAGING.actions.savefile,
+            name: filename,
+            data: fileStr,
+        }, '*');
+    }
+
+    /**
+     * Send a hex to the host controller to flash the device.
+     *
+     * @param {string} filename - File name for the hex to flash into the device.
+     * @param {string} hexStr  - Hex string to flash to the device.
+     */
+    function hostFlashHex(filename, hexStr) {
+        hostPostMessage({
+            type: CONTROLLER_MESSAGING.type,
+            action: CONTROLLER_MESSAGING.actions.flashhex,
+            name: filename,
+            data: hexStr,
+        }, '*');
+    }
+
```

Comment:
I'm wondering if flashhex and savefile should include the unmodified project name, so the app knows what is is and can decide what to call the file. And if savefile should send the "main.py" filename, when relevant, to allow the app to know that it is "main.py".

Is the project name stored in the hex?

Reply:
So instead of sending the filename generated by the Python editor (which is the filename users see when downloading the files) to pass the unmodified Script Name?

![image](https://user-images.githubusercontent.com/29712657/109045966-1f920000-76cc-11eb-97a0-df15a1a7beac.png)

Reply:
> which is the filename users see when downloading the files

The app should be in control of downloading files.

My first thought was about losing the script name when round tripping a hex. I've just realised that the website already suffers from this problem. Maybe I am missing a reason why it's best to use the filename when reloading a hex.

If preserving the script name is preferable, it would be best to have the it inside the hex, rather than exchange it separately.

Otherwise, I was thinking the messages could send the script name in addition to the file name, which provides the file type in the extension.

For Download, the app can use the provided filename or make it's own, and would be able to store the script name and send it back when round-tripping a hex.

For Flash, the script name would be more appropriate than the filename for any prompts.

In the 'loadhex' message the name is called filename but is only used to set the script name.

There may be a future case where the app would want to know it is "main.py". The app is capable of doing the same name substitution as the website before presenting it in a download or share dialogue.

I have a problem that the iOS app adds a number to the filename to cope with name clashes. Saving a modified "script" creates "script 2", the third modification comes back as "script_2", the fourth becomes "script_2 2".

Losing or modifying information in the messages complicates the situation.

Reply:
>>which is the filename users see when downloading the files

>The app should be in control of downloading files.

Right, I meant that's the name for the filename downloaded by the Python Editor on desktop. So it doesn't download main.py it downloads microbit_program.py. I think we are talking about the same thing, but just make sure, in the current version of the protocol the user script is sent with the filename `microbit_program.py` and not `main.py`.

> My first thought was about losing the script name when round tripping a hex. I've just realised that the website already suffers from this problem. Maybe I am missing a reason why it's best to use the filename when reloading a hex.

Roundtripping a hex file should maintain the filename, but unfortunately it does edit the "script name" as it replaces spaces with underscores when downloading the file and doesn't replace them back in the "script name" field when loading a hex. So I guess technically it is not exactly the same name in the editor, but it should be the same filename in the user filesystem.

With the current version of this PR a hex file in the app sends the hex string and hex filename to the editor, the editor adds the filename into the "script name" field (minus the extension), and then when clicking the editor download button, the editor sends to the app the filename derived from the "script name" (replaces spaces with underscores and adds the hex extension) together with the hex string. So it should result in the same filename after the round trip. If it doesn't then it's a bug we need to fix.

> If preserving the script name is preferable, it would be best to have the it inside the hex, rather than exchange it separately.
>Otherwise, I was thinking the messages could send the script name in addition to the file name, which provides the file type in the extension.

I guess the current script name changes has been present since the creating of this editor, so it's probably fine to continue like this. If anything I would probably update the editor to replace underscores with spaces when loading a hex file.

> I have a problem that the iOS app adds a number to the filename to cope with name clashes. Saving a modified "script" creates "script 2", the third modification comes back as "script_2", the fourth becomes "script_2 2".

Right, if a user has "script", "script 2", and "script 3" in the app (or desktop) files, then loading those files into the editor would result in having the extra numbers in the Script Name field as well. I think for the Python Editor on desktop this has been the pattern from the beginning, the script name has been directly tied to the filename, so I wouldn't change it there at this point.

However it sounds like the Script Name would be useful to the app, even if it is just for that name to be shown when flashing, so let's add it. 👍

Reply:
> in the current version of the protocol the user script is sent with the filename microbit_program.py and not main.py.

It would be better to send the script name and "main.py".

> So it should result in the same filename after the round trip.

It doesn't if the filename contains spaces.

My confusion is that I was expecting we would want the Python script name to behave like the MakeCode project name. MakeCode preserves the project name even if the hex file is renamed.

If you're happy that the script name is just the file name, then I guess we don't need both in the messages. I expect I can find a way round the app's file numbering problem.

The filename passed in the messages is not safe. I guess it's fixed in the browser by the website's "save as" library code. So the app needs to check and fix it.

Reply:
> If you're happy that the script name is just the file name, then I guess we don't need both in the messages. I expect I can find a way round the app's file numbering problem.

I think i'm happy with the current script name and file name relationship. I think the part that probably bothered me before was if the app said something like "flashing microbit_program (2)", but to be fair if that is what it says in the Script Name field, then that's what it is, but it would be fine as well if it got "cleaned" by the app.

> The filename passed in the messages is not safe. I guess it's fixed in the browser by the website's "save as" library code. So the app needs to check and fix it.

Yeah, I think "safeName" is a bit misleading in this case. The simple space to underscore change was added at the beginning and we never really changed it. I think you are right that any unsafe character was likely filtered by the browser.
At this point I don't think is worth updating that in the Python Editor, but if you need to filter the name anyway  for the app and prefer to do that from the editor side, then a PR is always welcome.